### PR TITLE
DEP: Add 'python_requires' to setup.py to drop 3.4 support

### LINKED ISSUE #20697: pip install panadas fails on numpy version
#### Code Sample, a copy-pastable example if possible

```
C:\Users\nickdmax
λ pip install pandas
Collecting pandas
  Downloading https://files.pythonhosted.org/packages/08/01/803834bc8a4e708aedebb133095a88a4dad9f45bbaf5ad777d2bea543c7e/pandas-0.22.0.tar.gz (11.3MB)
    100% |################################| 11.3MB 315kB/s
  Could not find a version that satisfies the requirement numpy==1.9.3 (from versions: 1.10.4, 1.11.0, 1.11.1rc1, 1.11.1, 1.11.2rc1, 1.11.2, 1.11.3, 1.12.0b1, 1.12.0rc1, 1.12.0rc2, 1.12.0, 1.12.1rc1, 1.12.1, 1.13.0rc1, 1.13.0rc2, 1.13.0, 1.13.1, 1.13.3, 1.14.0rc1, 1.14.0, 1.14.1, 1.14.2)
No matching distribution found for numpy==1.9.3

C:\Users\nickdmax
λ pip list
...
numpy               1.14.2
...
```

#### Problem description
I was unable to upgrade pandas (same error as above). I don't think I have ever had numpy 1.9.3 installed. Since there was a numpy directory under the current pandas directory I did an uninstall "pip uninstall pandas" to attempt to reinstall. 

#### Output of ``pip freeze``

<details>
vattrs==17.2.0                       
Automat==0.6.0                      
backcall==0.1.0                     
backports-abc==0.5                  
bleach==2.1.3                       
certifi==2017.4.17                  
chardet==3.0.4                      
colorama==0.3.9                     
constantly==15.1.0                  
cssselect==1.0.1                    
cycler==0.10.0                      
decorator==4.2.1                    
entrypoints==0.2.3                  
html5lib==1.0.1                     
hyperlink==17.3.0                   
idna==2.5                           
incremental==17.5.0                 
ipykernel==4.8.2                    
ipython==6.3.1                      
ipython-genutils==0.2.0             
ipywidgets==7.2.1                   
jedi==0.11.1                        
Jinja2==2.10                        
jsonschema==2.6.0                   
jupyter==1.0.0                      
jupyter-client==5.2.3               
jupyter-console==5.2.0              
jupyter-core==4.4.0                 
kiwisolver==1.0.1                   
MarkupSafe==1.0                     
matplotlib==2.2.2                   
mistune==0.8.3                      
nbconvert==5.3.1                    
nbformat==4.4.0                     
notebook==5.0.0                     
numpy==1.14.2                       
pandocfilters==1.4.2                
parso==0.1.1                        
pickleshare==0.7.4                  
prompt-toolkit==1.0.15              
PyDispatcher==2.0.5                 
Pygments==2.2.0                     
PyParse==1.1.7                      
pyparsing==2.2.0                    
python-dateutil==2.7.2              
pytz==2018.4                        
pywinpty==0.5.1                     
pyzmq==17.0.0                       
qtconsole==4.3.1                    
requests==2.18.2                    
Rx==1.5.9                           
scipy==0.19.1                       
simplegeneric==0.8.1                
six==1.11.0                         
testpath==0.3.1                     
tornado==5.0.2                      
traitlets==4.3.2                    
typing==3.6.4                       
urllib3==1.22                       
virtualenv==15.1.0                  
w3lib==1.17.0                       
wcwidth==0.1.7                      
webencodings==0.5.1                 
widgetsnbextension==3.2.1           
win-unicode-console==0.5            
zope.interface==4.4.2               
</details>

pip version 10.0.0

``python -c "import numpy; print(numpy.version.version)"``
1.14.2

### PR DESCRIPTION
**Pull request by djhoese**
- [x] closes #20697 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is a small change to the setup.py that *should* stop newer versions of pandas from being downloaded in environments with older/unsupported versions of python. I followed the instructions here: https://packaging.python.org/guides/dropping-older-python-versions/

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -748,4 +748,5 @@ def pxd(name):
       long_description=LONG_DESCRIPTION,
       classifiers=CLASSIFIERS,
       platforms='any',
+      python_requires='>2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*',
```
Comment (jreback) [inline setup.py]: hmm, there has to be a better way to write this, so signify *only* 2.7 or >=3.5

Comment (davidh-ssec) [reply setup.py]: I was following numpy (https://github.com/numpy/numpy/blob/master/setup.py#L368) and the link I mentioned.

Comment (davidh-ssec) [reply setup.py]: https://packaging.python.org/tutorials/distributing-packages/?highlight=python_requires#python-requires

Comment (davidh-ssec) [reply setup.py]: I agree that it should be simpler but wasn't sure what conditionals were allowed.

Comment (davidh-ssec) [reply setup.py]: Oh, I do notice a big mistake. It should be `>=2.7` not `>2.7`. Researching to see if there is a shorter way.

Comment (davidh-ssec) [reply setup.py]:
```
In [15]: from pip.utils.packaging import specifiers, version

In [16]: version.parse('3.5.0') in specifiers.SpecifierSet('==2.7,>=3.5.*')
Out[16]: False
```

Comment (jreback) [reply setup.py]: I *think* you can do ``>=3.5`` ?

**Comment by jreback** [comment]
cc @TomAugspurger

**Comment by davidh-ssec** [comment]
Moving the discussion outside of the commits in case more changes are done. Simple test for a simpler `python_requires`:

```
In [15]: from pip.utils.packaging import specifiers, version

In [16]: version.parse('3.5.0') in specifiers.SpecifierSet('==2.7,>=3.5.*')
Out[16]: False

# later I did:
In [2]: version.parse('3.5.0') in specifiers.SpecifierSet('==2.7,>=3.5')
Out[2]: False

```

**Comment by davidh-ssec** [comment]
I read the source for `SpecifierSet` and `Specifier`. The condition is basically `all(my_version in specifier for specifier in specifiers)` where the `__contains__` is doing the individual condition for each version specifier (`my_version == 2.7 and my_version >=3.5`).

I guess you'll just have to drop python 2.7 sooner /s

**Review by TomAugspurger — APPROVED** [review]

**Comment by davidh-ssec** [comment]
@jreback You want the setuptools version in `setup_requires`? Or is there somewhere else it is currently specified?

**Comment by davidh-ssec** [comment]
Oh you meant set the min version in install.rst and update the deps in the 0.23 whatsnew. For the deps do you want it added to the min version table of deps?

**Comment by jreback** [comment]
yes + in

```
(pandas) bash-3.2$ grep setuptools ci/*
ci/environment-dev.yaml:  - setuptools>=3.3
ci/requirements_dev.txt:setuptools>=3.3
```
 (just update these)

**Comment by jreback** [comment]
thanks @davidh-ssec

**Comment by TomAugspurger** [comment]
We’re hoping to do a release in about 1 week.

________________________________

**Comment by jorisvandenbossche** [comment]
Should we remove our `pyproject.toml` alltogether for upcoming 0.23.0 until support in pip is more mature? It seems that also the varying numpy version depending on the python version is not yet supported (cfr https://github.com/scipy/scipy/issues/8734)

**Pull request merged**